File: src/web/zigdom.zig
// https://github.com/shritesh/zig-wasm-dom
extern "document" fn query_selector(selector_ptr: [*]const u8, selector_len: usize) usize;
extern "document" fn create_element(tag_name_ptr: [*]const u8, tag_name_len: usize) usize;
extern "document" fn create_text_node(data_ptr: [*]const u8, data_len: usize) usize;
extern "element" fn set_attribute(element_id: usize, name_ptr: [*]const u8, name_len: usize, value_ptr: [*]const u8, value_len: usize) void;
extern "element" fn get_attribute(element_id: usize, name_ptr: [*]const u8, name_len: usize, value_ptr: *[*]u8, value_len: *usize) bool;
extern "event_target" fn add_event_listener(event_target_id: usize, event_ptr: [*]const u8, event_len: usize, event_id: usize) void;
extern "window" fn alert(msg_ptr: [*]const u8, msg_len: usize) void;
extern "window" fn console_log(msg_ptr: [*]const u8, msg_len: usize) void;
extern "node" fn append_child(node_id: usize, child_id: usize) usize;
extern "zig" fn release_object(object_id: usize) void;
extern "code" fn load_stack(stack_ptr: [*]const u8, stack_len: usize, stack_number: usize) void;

const assembler = @import("assembler");
const Cpu = @import("cpu").Cpu;

var cpu: Cpu = undefined;

fn log(msg: []const u8) void {
    console_log(msg.ptr, msg.len);
}

const std = @import("std");

export fn add(a: i32, b: i32) i32 {
    return a + b;
}

const eventId = enum(usize) {
    Submit,
    Clear,
};

var textarea_node: usize = undefined;

fn launch() !void {
    const body_selector = "body";
    const body_node = query_selector(body_selector, body_selector.len);
    defer release_object(body_node);

    if (body_node == 0) {
        return error.QuerySelectorError;
    }
    const input_tag_name = "div";
    const input_tag_node = create_element(input_tag_name, input_tag_name.len);
    defer release_object(input_tag_node);

    if (input_tag_node == 0) {
        return error.CreateElementError;
    }

    const attribute_name = "innerText";
    const attribute_value = "hello world";

    set_attribute(input_tag_node, attribute_name.ptr, attribute_name.len, attribute_value.ptr, attribute_value.len);

    const attached_input_node = append_child(body_node, input_tag_node);
    defer release_object(attached_input_node);

    if (attached_input_node == 0) {
        return error.AppendChildError;
    }

    const textarea_selector = "#code";
    textarea_node = query_selector(textarea_selector.ptr, textarea_selector.len);
    const textarea_value_name = "value";
    var textarea_value: [*]u8 = undefined;
    var textarea_len: usize = undefined;
    const success = get_attribute(textarea_node, textarea_value_name.ptr, textarea_value_name.len, &textarea_value, &textarea_len);

    if (success) {
        const result = textarea_value[0..textarea_len];
        defer std.heap.page_allocator.free(result);

        console_log(result.ptr, result.len);
    }

}

export fn load_code(code_ptr: [*]u8, code_len: usize) void {
    // const result = code_ptr[0..code_len];
    // defer std.heap.page_allocator.free(result);

    // log(result.ptr, result.len);
    console_log(code_ptr, code_len);

    const buf = code_ptr[0..code_len];
    var fbs = std.io.fixedBufferStream(buf);
    const reader = fbs.reader();

    const allocator = std.heap.wasm_allocator;

    var parser = assembler.tokenizer.Parser(@TypeOf(reader)).init(reader, allocator);
    defer parser.deinit();

    const tokens = parser.parse() catch {
        log("parser failed");
        unreachable;
    };

    var ast = assembler.ast.AST.init(tokens, allocator) catch {
        log("ast failed");
        unreachable;
    };
    defer ast.deinit();

    var outbuf: [1024]u8 = undefined;
    var fbout = std.io.fixedBufferStream(&outbuf);

    ast.writeCode(fbout.writer()) catch {
        log("ast -> machine code failed");
        unreachable;
    };

    fbout.reset();

    cpu = Cpu.init(fbout.reader()) catch {
        // TODO
        log("failed to init CPU");
        unreachable;
    };

    cpu.step() catch {
        log("failed to run CPU");
        unreachable;
    };

    cpu.step() catch {
        log("failed to run CPU");
        unreachable;
    };

    load_stack(&cpu.s0.data, 16, 0);
    load_stack(&cpu.s1.data, 16, 1);
    log(&cpu.s0.data);
    log(&cpu.s1.data);
}

export fn launch_export() bool {
    launch() catch {
        return false;
    };
    return true;
}

export fn _wasm_alloc(len: usize) u32 {
    const buf = std.heap.page_allocator.alloc(u8, len) catch {
        return 0;
    };
    return @intFromPtr(buf.ptr);
}

File: zigdom.js
var stackElements = [[], []];
var currentCode = "";

// init stacks
const st = [document.getElementById("stack1"), document.getElementById("stack2")];

for (let i = 0; i < 2; i++) {
  for (let j = 0; j < 16; j++) {
    const div = document.createElement("div");
    div.classList.add("stack-element");
    stackElements[i].push(div)
    st[i].appendChild(div);
  }
}

const getString = function(ptr, len) {
  const slice = zigdom.exports.memory.buffer.slice(ptr, ptr + len);
  const textDecoder = new TextDecoder();
  return textDecoder.decode(slice);
};

const getUint8Array = function(ptr, len) {
  return new Uint8Array(zigdom.exports.memory.buffer, ptr, len);
}

const pushObject = function(object) {
  return zigdom.objects.push(object);
};

const getObject = function(objId) {
  return zigdom.objects[objId - 1];
};

const dispatch = function(eventId) {
  return function() {
    zigdom.exports.dispatchEvent(eventId);
  };
};

const elementSetAttribute = function(
  node_id,
  name_ptr,
  name_len,
  value_ptr,
  value_len
) {
  const node = getObject(node_id);
  const attribute_name = getString(name_ptr, name_len);
  const value = getString(value_ptr, value_len);
  node[attribute_name] = value;
};

const compileBtn = document.getElementById("compile");

compileBtn.addEventListener("click", function(event) {
  const area = document.getElementById("code");
  const text = area.value;
  currentCode = text;

  const textEncoder = new TextEncoder();
  const resultArray = textEncoder.encode(text);
  const len = resultArray.length;

  if (len === 0) {
    return false;
  }

  const ptr = zigdom.exports._wasm_alloc(len);
  if (ptr === 0) {
    throw "Cannot allocate memory";
  }

  // write the array to the memory
  const mem_result = new DataView(zigdom.exports.memory.buffer, ptr, len);
  for (let i = 0; i < len; ++i) {
    mem_result.setUint8(i, resultArray[i], true);
  }

  zigdom.exports.load_code(ptr, len);
})

function hasCompiled() {
  return currentCode === document.getElementById("code").value;
}

const stepBtn = document.getElementById("step");
stepBtn.addEventListener("click", function(event) {
  if (!hasCompiled()) {
    compileBtn.click();
  }
  zigdom.exports.step_cpu();
});

const runBtn = document.getElementById("run");
runBtn.addEventListener("click", function(event) {
  if (!hasCompiled()) {
    compileBtn.click();
  }
  zigdom.exports.run_cpu();
});

const loadStack = function(
  stack_ptr,
  stack_len,
  stack_number,
) {
  const stack = getUint8Array(stack_ptr, stack_len);
  for (let i = 0; i < 16; i++) {
    let pos = stack[i];
    if (i >= stack_len) pos = " ";
    stackElements[stack_number][i].innerText = pos;
  }
}

const elementGetAttribute = function(
  node_id,
  name_ptr,
  name_len,
  result_address_ptr,
  result_address_len_ptr
) {
  const node = getObject(node_id);
  const attribute_name = getString(name_ptr, name_len);
  const result = node[attribute_name];
  // convert result into Uint8Array
  const textEncoder = new TextEncoder();
  const resultArray = textEncoder.encode(result);
  var len = resultArray.length;

  if (len === 0) {
    return false;
  }

  // allocate required number of bytes
  const ptr = zigdom.exports._wasm_alloc(len);
  if (ptr === 0) {
    throw "Cannot allocate memory";
  }

  // write the array to the memory
  const mem_result = new DataView(zigdom.exports.memory.buffer, ptr, len);
  for (let i = 0; i < len; ++i) {
    mem_result.setUint8(i, resultArray[i], true);
  }

  // write the address of the result array to result_address_ptr
  const mem_result_address = new DataView(
    zigdom.exports.memory.buffer,
    result_address_ptr,
    32 / 8
  );
  mem_result_address.setUint32(0, ptr, true);

  //write the size of the result array to result_address_ptr_len_ptr
  const mem_result_address_len = new DataView(
    zigdom.exports.memory.buffer,
    result_address_len_ptr,
    32 / 8
  );
  mem_result_address_len.setUint32(0, len, true);

  // return if success? (optional)
  return true;
};
const eventTargetAddEventListener = function(
  objId,
  event_ptr,
  event_len,
  eventId
) {
  const node = getObject(objId);
  const ev = getString(event_ptr, event_len);
  node.addEventListener(ev, dispatch(eventId));
};

const documentQuerySelector = function(selector_ptr, selector_len) {
  const selector = getString(selector_ptr, selector_len);
  return pushObject(document.querySelector(selector));
};

const documentCreateElement = function(tag_name_ptr, tag_name_len) {
  const tag_name = getString(tag_name_ptr, tag_name_len);
  return pushObject(document.createElement(tag_name));
};

const documentCreateTextNode = function(data_ptr, data_len) {
  data = getString(data_ptr, data_len);
  return pushObject(document.createTextNode(data));
};

const nodeAppendChild = function(node_id, child_id) {
  const node = getObject(node_id);
  const child = getObject(child_id);

  if (node === undefined || child === undefined) {
    return 0;
  }

  return pushObject(node.appendChild(child));
};

const windowAlert = function(msg_ptr, msg_len) {
  const msg = getString(msg_ptr, msg_len);
  alert(msg);
};

const zigReleaseObject = function(object_id) {
  zigdom.objects[object_id - 1] = undefined;
};

const launch = function(result) {
  zigdom.exports = result.instance.exports;
  if (!zigdom.exports.launch_export()) {
    throw "Launch Error";
  }
};

const console_log = function(msg_ptr, msg_len) {
  const msg = getString(msg_ptr, msg_len);
  console.log(msg);
}

var zigdom = {
  objects: [],
  imports: {
    document: {
      query_selector: documentQuerySelector,
      create_element: documentCreateElement,
      create_text_node: documentCreateTextNode
    },
    element: {
      set_attribute: elementSetAttribute,
      get_attribute: elementGetAttribute
    },
    event_target: {
      add_event_listener: eventTargetAddEventListener
    },
    node: {
      append_child: nodeAppendChild
    },
    window: {
      alert: windowAlert,
      console_log: console_log
    },
    zig: {
      release_object: zigReleaseObject
    },
    code: {
      load_stack: loadStack
    }
  },
  launch: launch,
  exports: undefined
};